build/native.rs

```diff
@@ -33,35 +31,76 @@ const DEFAULT_SDK_DIR: &str = ".espressif";
 const DEFAULT_ESP_IDF_REPOSITORY: &str = "https://github.com/espressif/esp-idf.git";
 const DEFAULT_ESP_IDF_VERSION: &str = "v4.3";
 
-fn esp_idf_version() -> git::Ref {
-    let version =
-        env::var(ESP_IDF_VERSION_VAR).unwrap_or_else(|_| DEFAULT_ESP_IDF_VERSION.to_owned());
-    let version = version.trim();
-    assert!(
-        !version.is_empty(),
-        "${} (='{}') must contain a valid version",
-        ESP_IDF_VERSION_VAR,
-        version
-    );
+const CARGO_CMAKE_BUILD_ACTIVE_VAR: &str = "CARGO_CMAKE_BUILD_ACTIVE";
+const CARGO_CMAKE_BUILD_INCLUDES_VAR: &str = "CARGO_CMAKE_BUILD_INCLUDES";
+const CARGO_CMAKE_BUILD_LINK_LIBRARIES_VAR: &str = "CARGO_CMAKE_BUILD_LINK_LIBRARIES";
+const CARGO_CMAKE_BUILD_COMPILER_VAR: &str = "CARGO_CMAKE_BUILD_COMPILER";
+const CARGO_CMAKE_BUILD_SDKCONFIG_VAR: &str = "CARGO_CMAKE_BUILD_SDKCONFIG";
 
-    match version.split_once(':') {
-        Some(("commit", c)) => git::Ref::Commit(c.to_owned()),
-        Some(("tag", t)) => git::Ref::Tag(t.to_owned()),
-        Some(("branch", b)) => git::Ref::Branch(b.to_owned()),
-        _ => match version.chars().next() {
-            Some(c) if c.is_ascii_digit() => git::Ref::Tag("v".to_owned() + version),
-            Some('v') if version.len() > 1 && version.chars().nth(1).unwrap().is_ascii_digit() => {
-                git::Ref::Tag(version.to_owned())
-            }
-            Some(_) => git::Ref::Branch(version.to_owned()),
-            _ => unreachable!(),
-        },
+pub fn build() -> Result<EspIdfBuildOutput> {
+    if env::var(CARGO_CMAKE_BUILD_ACTIVE_VAR).is_ok()
+        || env::var(CARGO_CMAKE_BUILD_INCLUDES_VAR).is_ok()
+    {
+        build_cmake_first()
+    } else {
+        build_cargo_first()
     }
 }
 
-pub(crate) fn main() -> Result<EspIdfBuildOutput<impl Iterator<Item = (String, kconfig::Value)>>> {
-    // TODO: Implement support for CMake-first builds in a similar way as to how cargo-pio supports PIO-first builds
+fn build_cmake_first() -> Result<EspIdfBuildOutput> {
+    const IDF_LIB_PREFIX: &str = "__idf_";
```

Comment:
This is really cmake specific (the `__<name>_` part). I looked at this in the past and didn't find where the separation of target namespaces is documented and I assumed that this is unstable. So I think maybe removing this prefix should be done in the cmake script itself.

But this isn't really that critical: maybe add a TODO.

Reply:
Even if this is not just CMake-specific but in fact ESP-IDF+CMake-specific, I don't think that's a big problem, because we are *assuming* that we are being built with CMake? Even the env-vars are named with a `CMAKE_` prefix?

Reply:
I'm saying that this namespace separation may be cmake version-specific because I've not found it documented anywhere.

Reply:
This unfortunately is an internal detail of IDF build system. For each component `name`, we define an alias target `idf::name`. Alias targets are visible only inside CMake, so for the library targets visible outside (e.g. in the generated Ninja build scripts) we use `__idf_name` pattern, see [here](https://github.com/espressif/esp-idf/blob/3e370c4296247b349aa3b9a0076c05b9946d47dc/tools/cmake/component.cmake#L174-L181) and [here](https://github.com/espressif/esp-idf/blob/3e370c4296247b349aa3b9a0076c05b9946d47dc/tools/cmake/component.cmake#L47-L49).

If you think this is an issue, we can try to export some variable from the build system which would contain only the plain list of components, without this prefix.

The reason it is a bit difficult, though, is that in general the `LINK_LIBRARIES` list might include libraries which aren't IDF components. For example, the user may add a custom library in their top-level CMake file, and the name of such library target will not have the same pattern.

Reply:
Nope, that's good news actually. You are essentially saying that (a) `__idf_` is specific for the **ESP-IDF** CMake build and (b) it is NOT CMake *version* specific. So we can rely on it not changing when CMake itself is upgraded to a new version.

Reply:
Ah, that's good to know. I thought that's how cmake exports aliased (or namespaced) targets, but this isn't the case. So then I agree, having this logic here is no problem.

Reply:
Assuming ESP-IDF doesn't change this internal detail, I think this is okay. (I can't think of a reason this would be changed anytime soon)